contracts/global-extensions/GlobalOptimisticAuctionRebalanceExtension.sol

```diff
@@ -0,0 +1,328 @@
+    mapping(bytes32 => bytes32) public assertionIds; // Maps proposal hashes to assertionIds.
+    mapping(bytes32 => bytes32) public proposalHashes; // Maps assertionIds to proposal hashes.
+    mapping (bytes32=>ISetToken) public assertedProducts; // Maps assertionIds to ISetToken
+
+    // Keys for assertion claim data.
+    bytes public constant PROPOSAL_HASH_KEY = "proposalHash";
+    bytes public constant RULES_KEY = "rulesIPFSHash";
+
+
+    /* ============ Constructor ============ */
+
+    constructor(AuctionExtensionParams memory _auctionParams) public GlobalAuctionRebalanceExtension(_auctionParams.managerCore, _auctionParams.auctionModule) {
+    
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+    * @dev OPERATOR ONLY: sets product settings for a given quote asset
+    * @param _product Address of the SetToken to set rules and settings for.
+    * @param _optimisticParams OptimisticRebalanceParams struct containing optimistic rebalance parameters.
+    * @param _rulesHash bytes32 containing the ipfs hash rules for the product.
+    */
+    function setProductSettings(
+        ISetToken _product,
+        OptimisticRebalanceParams memory _optimisticParams,
+        bytes32 _rulesHash
+    )
+        external
+        onlyOperator(_product)
+    {
+        productSettings[_product] = ProductSettings({
+            optimisticParams: _optimisticParams,
+            rulesHash: _rulesHash
+        });
+    }
+
+     /**
+    * @dev OPERATOR ONLY: sets product settings for a given quote asset
+     * @param _quoteAsset                   ERC20 token used as the quote asset in auctions.
+     * @param _oldComponents                Addresses of existing components in the SetToken.
+     * @param _newComponents                Addresses of new components to be added.
+     * @param _newComponentsAuctionParams   AuctionExecutionParams for new components, indexed corresponding to _newComponents.
+     * @param _oldComponentsAuctionParams   AuctionExecutionParams for existing components, indexed corresponding to
+     *                                      the current component positions. Set to 0 for components being removed.
+     * @param _shouldLockSetToken           Indicates if the rebalance should lock the SetToken.
+     * @param _rebalanceDuration            Duration of the rebalance in seconds.
+     * @param _positionMultiplier           Position multiplier at the time target units were calculated.
+    */
+    function proposeRebalance(
+        ISetToken _setToken,
+        IERC20 _quoteAsset,
+        address[] memory _oldComponents,
+        address[] memory _newComponents,
+        AuctionExecutionParams[] memory _newComponentsAuctionParams,
+        AuctionExecutionParams[] memory _oldComponentsAuctionParams,
+        bool _shouldLockSetToken,
+        uint256 _rebalanceDuration,
+        uint256 _positionMultiplier
```

Comment:
If we want this to be called by anyone we might want to add an argument to pass the maximum bond that the caller is willing to put up. Right now I think technically if someone sends this transaction, we could frontrun him change the bond settings to something crazy and potentially lock up all of his tokens. (if the sender has high enough approval).
Obviously not something we are going to do but might be a good addition to give potential callers of this method extra control over their funds.

Reply:
Not implemented for now as it is outside of the scope to limit the bad actions the operator could take.